### ISSUE TITLE
Add capability to limit cache size 

### ISSUE THREAD

**Issue by jvoegele** [cabol/nebulex#53]
In order to prevent memory usage from growing out of control, it would be nice to be able to limit the size of Nebulex caches. If the size limit has been reached, then a new item cannot be added to the cache without evicting some other item from the cache first. Eviction policy could be LRU, LFU, or perhaps based on generation.

Ideally, the size limitation would be in terms of the actual amount of memory occupied by the cache values, but if that is difficult to determine efficiently, then the number of keys in the cache could be used as a reasonable surrogate.

**Comment by cabol** [comment]
Agreed, that would be a nice feature. I will investigate what would be a good way to implement it (the size limitation can be done in terms of the actual amount of memory occupied in bytes or either based on the number of keys, maybe both?)

**Comment by yuraxdrumz** [comment]
Hey, 
It is a very nice feature to have and to create an initial implementation, I referenced our beloved Redis eviction policy. Maybe we can start with an LRU cache like in Redis. 
The flow goes like this:
1. expire is called on some key or if memory limit reached or if some upper threshold we decide
2. sample N keys (Redis uses 5 usually) to populate a pool of best keys to evict of M keys (16 in Redis)
3. every key from N is added to evict pool if it has higher old access time (pool will be access time sorted).
4. When pool is full, evict best key, but do not delete it (Redis does not delete it, I guess for performance reasons?)
5. When all entries in the pool are already deleted, a.k.a, do not exist, start populating pool again until full.

Inserting can work like this:
- find first empty bucket from pool, lets call it K.
- if the first element is not empty and its access time is better than our candidate K, continue
- else if we have an empty bucket and key is null, we simply insert key K.
-  else we insert in the middle, so our candidate K points to first element greater than K and we shift all the elements from K to the right
- If no free space on right, K--, shift all elements to left, discarding lowest idle key.

I know its Redis's, but its battle tested and a great start for an LRU cache size

**Comment by cabol** [comment]
Hey, thanks for the detailed suggestion, sounds very good 👍 . At this stage is nice to receive proposals in order to have a better idea of what path to follow. I will definitely review it more in details and try to start working on this very soon!

On the other hand, independent of implementation, I was thinking to add it in a pluggable and/or configurable way, taking advantage of the hooks plug the eviction mechanism to control and protect the memory and make it configurable via `:max_size` parameter, if it is not set, just skip or avoid to run all this.

Anyways, thanks again, and remember any contribution is welcomed 😄

**Comment by yuraxdrumz** [comment]
Sounds great, the only thing I am worried about is performance, but that can be easily tested with benchee. I will try and find some time this week and implement a proof of concept for this.

**Comment by cabol** [comment]
> Sounds great, the only thing I am worried about is performance

Indeed, that's why I was thinking of something like, with a post_hook send events (detailed event with the command info) to a GenServer (using `GenServer.cast`), and perform the logic there.

> I will try and find some time this week and implement a proof of concept for this.

That would be awesome!

**Comment by cabol** [comment]
I implemented a simple solution for now (we can add more improvements later). Following the same idea of the generational cache, I created two options to control/limit the cache size:

* `:generation_size` - Max size in bytes for a cache generation. If this option is set and the configured value is reached, a new generation is created so the oldest is deleted and force releasing memory space. If it is not set (`nil`), the cleanup check to release memory is not performed (the default).

* `:gc_cleanup_interval` - The number of writes needed to run the cleanup check. Once this value is reached and only if `generation_size` option is set, the cleanup check is performed. Defaults to `10`, so after 10 write operations the cleanup check is performed.

In the `Nebulex.Adapters.Local` after a write is executed, the cleanup check is performed. If `:generation_size` is not set, nothing happens, and the behavior is the previous one (no memory check). Bit if it is set, the cleanup process is triggered by means of a `GenServer.cast` in order to avoid affecting the performance too much. And the GC in a separate process performs the validation and the corresponding actions described in the options above.

I think for now it is good enough, this allows to limit the cache size (limiting the generation size) and prevent memory usage from growing out of control and re-use the "generational approach" to release memory.

Let me know your thoughts about it!
